### ISSUE TITLE
[Routing][i18n] Wrong PHP format example

### ISSUE THREAD

**Issue by fsevestre** [symfony/symfony-docs#10762]
Hello,

There are some issues in the PHP format example (https://github.com/symfony/symfony-docs/blame/e350b3efcfadbb26ab51768e5f6a2bcb07d9dc14/routing.rst#L260) which don't generate the same result as the YAML or the XML formats examples.

--------------------

```yaml
# config/routes/annotations.yaml

controllers:
    resource: ../../src/Controller/
    type: annotation
    prefix:
        en: '' # don't prefix URLs for English, the default locale
        nl: '/nl'
        it: '/it'
```

```php
// var/cache/dev/srcApp_KernelDevDebugContainerUrlMatcher

<?php

use Symfony\Component\Routing\Matcher\Dumper\PhpMatcherTrait;
use Symfony\Component\Routing\RequestContext;

/**
 * This class has been auto-generated
 * by the Symfony Routing Component.
 */
class srcApp_KernelDevDebugContainerUrlMatcher extends Symfony\Bundle\FrameworkBundle\Routing\RedirectableUrlMatcher
{
    use PhpMatcherTrait;

    public function __construct(RequestContext $context)
    {
        $this->context = $context;
        $this->staticRoutes = array(
            '/' => array(array(array('_route' => 'app_default_index', '_controller' => 'App\\Controller\\DefaultController::index', '_locale' => 'en'), null, null, null, false, null)),
            '/nl' => array(array(array('_route' => 'app_default_index', '_controller' => 'App\\Controller\\DefaultController::index', '_locale' => 'nl'), null, null, null, true, null)),
            '/it' => array(array(array('_route' => 'app_default_index', '_controller' => 'App\\Controller\\DefaultController::index', '_locale' => 'it'), null, null, null, true, null)),
            '/test' => array(array(array('_route' => 'app_default_index2', '_controller' => 'App\\Controller\\DefaultController::index2', '_locale' => 'en'), null, null, null, false, null)),
            '/nl/test' => array(array(array('_route' => 'app_default_index2', '_controller' => 'App\\Controller\\DefaultController::index2', '_locale' => 'nl'), null, null, null, false, null)),
            '/it/test' => array(array(array('_route' => 'app_default_index2', '_controller' => 'App\\Controller\\DefaultController::index2', '_locale' => 'it'), null, null, null, false, null)),
        );
        $this->regexpList = array(
            0 => '{^(?'
                    .'|/_error/(\\d+)(?:\\.([^/]++))?(*:35)'
                .')(?:/?)$}sDu',
        );
        $this->dynamicRoutes = array(
            35 => array(array(array('_route' => '_twig_error_test', '_controller' => 'twig.controller.preview_error::previewErrorPageAction', '_format' => 'html'), array('code', '_format'), null, null, false, null)),
        );
    }
}
```

```
php bin/console debug:router

 ----------------------- -------- -------- ------ -------------------------- 
  Name                    Method   Scheme   Host   Path                      
 ----------------------- -------- -------- ------ -------------------------- 
  app_default_index.en    ANY      ANY      ANY    /                         
  app_default_index.nl    ANY      ANY      ANY    /nl/                      
  app_default_index.it    ANY      ANY      ANY    /it/                      
  app_default_index2.en   ANY      ANY      ANY    /test                     
  app_default_index2.nl   ANY      ANY      ANY    /nl/test                  
  app_default_index2.it   ANY      ANY      ANY    /it/test                  
  _twig_error_test        ANY      ANY      ANY    /_error/{code}.{_format}  
 ----------------------- -------- -------- ------ --------------------------
```

--------------------

```php
// config/routes/annotations.php

<?php

use Symfony\Component\Routing\RouteCollection;

/** @var RouteCollection $routes */
$routes = $loader->import('../../src/Controller/', 'annotation');

$routes->addPrefix('/', array('_locale' => 'en'));
$routes->addPrefix('/nl', array('_locale' => 'nl'));
$routes->addPrefix('/it', array('_locale' => 'it'));

return $routes;
```

```php
// var/cache/dev/srcApp_KernelDevDebugContainerUrlMatcher

<?php

use Symfony\Component\Routing\Matcher\Dumper\PhpMatcherTrait;
use Symfony\Component\Routing\RequestContext;

/**
 * This class has been auto-generated
 * by the Symfony Routing Component.
 */
class srcApp_KernelDevDebugContainerUrlMatcher extends Symfony\Bundle\FrameworkBundle\Routing\RedirectableUrlMatcher
{
    use PhpMatcherTrait;

    public function __construct(RequestContext $context)
    {
        $this->context = $context;
        $this->staticRoutes = array(
            '/it/nl' => array(array(array('_route' => 'app_default_index', '_controller' => 'App\\Controller\\DefaultController::index', '_locale' => 'it'), null, null, null, true, null)),
            '/it/nl/test' => array(array(array('_route' => 'app_default_index2', '_controller' => 'App\\Controller\\DefaultController::index2', '_locale' => 'it'), null, null, null, false, null)),
        );
        $this->regexpList = array(
            0 => '{^(?'
                    .'|/_error/(\\d+)(?:\\.([^/]++))?(*:35)'
                .')(?:/?)$}sDu',
        );
        $this->dynamicRoutes = array(
            35 => array(array(array('_route' => '_twig_error_test', '_controller' => 'twig.controller.preview_error::previewErrorPageAction', '_format' => 'html'), array('code', '_format'), null, null, false, null)),
        );
    }
}
```

```
php bin/console debug:router

 -------------------- -------- -------- ------ -------------------------- 
  Name                 Method   Scheme   Host   Path                      
 -------------------- -------- -------- ------ -------------------------- 
  app_default_index    ANY      ANY      ANY    /it/nl/                   
  app_default_index2   ANY      ANY      ANY    /it/nl/test               
  _twig_error_test     ANY      ANY      ANY    /_error/{code}.{_format}  
 -------------------- -------- -------- ------ --------------------------
```

**Comment by xabbuh** [comment]
Can you please check if `$routes->addPrefix('', array('_locale' => 'en'));` would lead to the expected result?

**Comment by fsevestre** [comment]
It didn't change anything.

I managed to make it work, after digging in the [`YamlFileLoader`](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Routing/Loader/YamlFileLoader.php#L191)/[`XmlFileLoader`](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Routing/Loader/XmlFileLoader.php#L190):

```php
// config/routes/annotations.php

<?php

use Symfony\Component\Routing\RouteCollection;

/** @var RouteCollection $routes */
$routes = $loader->import('../src/Controller/', 'annotation');

$trailingSlashOnRoot = true;

$prefixes = [
    'en' => '',
    'nl' => '/nl',
    'it' => '/it',
];

foreach ($prefixes as $locale => $localePrefix) {
    $prefixes[$locale] = trim(trim($localePrefix), '/');
}

foreach ($routes->all() as $name => $route) {
    if (null === $locale = $route->getDefault('_locale')) {
        $routes->remove($name);

        foreach ($prefixes as $locale => $localePrefix) {
            $localizedRoute = clone $route;
            $localizedRoute->setDefault('_locale', $locale);
            $localizedRoute->setDefault('_canonical_route', $name);
            $localizedRoute->setPath($localePrefix.(!$trailingSlashOnRoot && '/' === $route->getPath() ? '' : $route->getPath()));
            $routes->add($name.'.'.$locale, $localizedRoute);
        }
    } elseif (!isset($prefixes[$locale])) {
        throw new \InvalidArgumentException(sprintf('Route "%s" with locale "%s" is missing a corresponding prefix when imported in "%s".', $name, $locale, $file));
    } else {
        $route->setPath($prefixes[$locale].(!$trailingSlashOnRoot && '/' === $route->getPath() ? '' : $route->getPath()));
        $routes->add($name, $route);
    }
}

return $routes;
```

Or simplify by:
```php
// config/routes/annotations.php

<?php

use Symfony\Component\Routing\Loader\Configurator\ImportConfigurator;
use Symfony\Component\Routing\RouteCollection;

/** @var RouteCollection $routes */
$routes = $loader->import('../src/Controller/', 'annotation');

$prefixes = [
    'en' => '',
    'nl' => '/nl',
    'it' => '/it',
];

$importConfigurator = new ImportConfigurator(new RouteCollection(), $routes);
$importConfigurator->prefix($prefixes);

return $routes;
```

But I don't what should really be the first argument of `ImportConfigurator` or simply if it is supposed to be use here :/

**Comment by javiereguiluz** [comment]
@fsevestre I'm a bit lost here. Do you think this could be a code bug instead of a doc problem? Thanks.

**Comment by fsevestre** [comment]
I we consider that the PHP format documentation is correct (by reading the doc code, it "seems" logical and correct to me), I would say it's more a code bug because the "prefix" system has different behavior depending on the routing format (YAML == XML != PHP).

**Comment by HeahDude** [comment]
I confirm the behavior is not implemented within a core `RouteCollection::addPrefix` call.
One should use the php configurator instead:
```php
<?php

namespace Symfony\Component\Routing\Loader\Configurator;

return function (RoutingConfigurator $routes) {
    $routes->import('../../src/Controller/', 'annotation')
        ->prefix([
            'en' => '',
            'nl' => '/nl',
            'it' => '/it',
        ])
    ;
};
```

We should close here as fixed by #11085, and maybe open an issue in Symfony core if we want to support it before documenting it.

**Comment by HeahDude** [comment]
Thank you @fsevestre for spotting this issue. As said before this is not a bug but a limitation because the feature as not been implemented at low level in routing core (i.e `RoutingCollection`).

Hopefully, this will be fixed by #11085 for which we already have a tracking issue.

Also, we both have left a solution here in case somebody encounter this same issue, so I'm going to close here.

Feel free to suggest a new feature in Symfony core (I suggest you open an issue first before creating a PR). The solution you found could easily live in a new `RouteCollection::setPrefixes(array $prefixes)` method or some kind 👍.

Thanks!
